Give batched messages wrong packet id.

**Describe the bug**
When I try to use Pulsar Client to send messages to MQTT consumers. the MQTT consumer will get an error as below. That caused by duplicated packet id.


### Error stacktrace

```java
com.hivemq.client.mqtt.exceptions.MqttSessionExpiredException: Session expired as connection was closed.

	at com.hivemq.client.internal.mqtt.MqttBlockingClient$MqttPublishes.handleError(MqttBlockingClient.java:402)
	at com.hivemq.client.internal.mqtt.MqttBlockingClient$MqttPublishes.receive(MqttBlockingClient.java:310)
	at io.streamnative.pulsar.handlers.mqtt.mqtt5.hivemq.base.MQTT5BatchMessageTest.testReceiveBatchMessageWithCorrectPacketId(MQTT5BatchMessageTest.java:45)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:54)
	at org.testng.internal.InvokeMethodRunnable.run(InvokeMethodRunnable.java:44)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: com.hivemq.client.mqtt.mqtt5.exceptions.Mqtt5DisconnectException: QoS 1 PUBLISH must not be resent during the same connection
```

### Received messages info

```java
22:43:04.736 [mock-pulsar-bk-OrderedExecutor-0-0:io.streamnative.pulsar.handlers.mqtt.support.MQTTConsumer@104] DEBUG io.streamnative.pulsar.handlers.mqtt.support.MQTTConsumer - [persistent://public/default/test-batch-message-1] [persistent://public/default/test-batch-message-1] [ffe25d9c-7d47-4335-82b1-5317986ea083] Send MQTT message MqttPublishMessage[fixedHeader=MqttFixedHeader[messageType=PUBLISH, isDup=false, qosLevel=AT_LEAST_ONCE, isRetain=false, remainingLength=0], variableHeader=MqttPublishVariableHeader[topicName=persistent://public/default/test-batch-message-1, packetId=1], payload=UnpooledSlicedByteBuf(ridx: 0, widx: 36, cap: 36/36, unwrapped: UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeDirectByteBuf(ridx: 99, widx: 99, cap: 8192))] to subscriber
22:43:04.737 [mock-pulsar-bk-OrderedExecutor-0-0:io.streamnative.pulsar.handlers.mqtt.support.MQTTConsumer@104] DEBUG io.streamnative.pulsar.handlers.mqtt.support.MQTTConsumer - [persistent://public/default/test-batch-message-1] [persistent://public/default/test-batch-message-1] [ffe25d9c-7d47-4335-82b1-5317986ea083] Send MQTT message MqttPublishMessage[fixedHeader=MqttFixedHeader[messageType=PUBLISH, isDup=false, qosLevel=AT_LEAST_ONCE, isRetain=false, remainingLength=0], variableHeader=MqttPublishVariableHeader[topicName=persistent://public/default/test-batch-message-1, packetId=2], payload=UnpooledSlicedByteBuf(ridx: 0, widx: 36, cap: 36/36, unwrapped: UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeDirectByteBuf(ridx: 233, widx: 233, cap: 8192))] to subscriber
22:43:04.737 [mock-pulsar-bk-OrderedExecutor-0-0:io.streamnative.pulsar.handlers.mqtt.support.MQTTConsumer@104] DEBUG io.streamnative.pulsar.handlers.mqtt.support.MQTTConsumer - [persistent://public/default/test-batch-message-1] [persistent://public/default/test-batch-message-1] [ffe25d9c-7d47-4335-82b1-5317986ea083] Send MQTT message MqttPublishMessage[fixedHeader=MqttFixedHeader[messageType=PUBLISH, isDup=false, qosLevel=AT_LEAST_ONCE, isRetain=false, remainingLength=0], variableHeader=MqttPublishVariableHeader[topicName=persistent://public/default/test-batch-message-1, packetId=2], payload=UnpooledSlicedByteBuf(ridx: 0, widx: 36, cap: 36/36, unwrapped: UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeDirectByteBuf(ridx: 233, widx: 233, cap: 8192))] to subscriber
22:43:04.737 [mock-pulsar-bk-OrderedExecutor-0-0:io.streamnative.pulsar.handlers.mqtt.support.MQTTConsumer@104] DEBUG io.streamnative.pulsar.handlers.mqtt.support.MQTTConsumer - [persistent://public/default/test-batch-message-1] [persistent://public/default/test-batch-message-1] [ffe25d9c-7d47-4335-82b1-5317986ea083] Send MQTT message MqttPublishMessage[fixedHeader=MqttFixedHeader[messageType=PUBLISH, isDup=false, qosLevel=AT_LEAST_ONCE, isRetain=false, remainingLength=0], variableHeader=MqttPublishVariableHeader[topicName=persistent://public/default/test-batch-message-1, packetId=2], payload=UnpooledSlicedByteBuf(ridx: 0, widx: 36, cap: 36/36, unwrapped: UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeDirectByteBuf(ridx: 233, widx: 233, cap: 8192))] to subscriber
22:43:04.737 [mock-pulsar-bk-OrderedExecutor-0-0:io.streamnative.pulsar.handlers.mqtt.support.MQTTConsumer@104] DEBUG io.streamnative.pulsar.handlers.mqtt.support.MQTTConsumer - [persistent://public/default/test-batch-message-1] [persistent://public/default/test-batch-message-1] [ffe25d9c-7d47-4335-82b1-5317986ea083] Send MQTT message MqttPublishMessage[fixedHeader=MqttFixedHeader[messageType=PUBLISH, isDup=false, qosLevel=AT_LEAST_ONCE, isRetain=false, remainingLength=0], variableHeader=MqttPublishVariableHeader[topicName=persistent://public/default/test-batch-message-1, packetId=2], payload=UnpooledSlicedByteBuf(ridx: 0, widx: 36, cap: 36/36, unwrapped: UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeDirectByteBuf(ridx: 233, widx: 233, cap: 8192))] to subscriber

```

--- a/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/MQTTConsumer.java
+++ b/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/MQTTConsumer.java
@@ -90,15 +90,14 @@ public class MQTTConsumer extends Consumer {
         ChannelPromise promise = cnx.ctx().newPromise();
         MESSAGE_PERMITS_UPDATER.addAndGet(this, -totalMessages);
         for (Entry entry : entries) {
-            int packetId = 0;
-            if (MqttQoS.AT_MOST_ONCE != qos) {
-                packetId = packetIdGenerator.nextPacketId();
-                outstandingPacketContainer.add(new OutstandingPacket(this, packetId, entry.getLedgerId(),
-                        entry.getEntryId()));
-            }
             String toConsumerTopicName = PulsarTopicUtils.getToConsumerTopicName(mqttTopicName, pulsarTopicName);
             List<MqttPublishMessage> messages = PulsarMessageConverter.toMqttMessages(toConsumerTopicName, entry,
-                    packetId, qos);
+                    packetIdGenerator, qos);
+            if (MqttQoS.AT_MOST_ONCE != qos) {
+                messages.stream().map(message -> new OutstandingPacket(this,
+                        message.variableHeader().packetId(), entry.getLedgerId(),
+                        entry.getEntryId())).forEach(outstandingPacketContainer::add);
+            }
             for (MqttPublishMessage msg : messages) {
                 if (log.isDebugEnabled()) {
                     log.debug("[{}] [{}] [{}] Send MQTT message {} to subscriber", pulsarTopicName,

--- a/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/utils/PulsarMessageConverter.java
+++ b/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/utils/PulsarMessageConverter.java
@@ -18,6 +18,7 @@ import io.netty.buffer.ByteBuf;
 import io.netty.handler.codec.mqtt.MqttPublishMessage;
 import io.netty.handler.codec.mqtt.MqttQoS;
 import io.netty.util.concurrent.FastThreadLocal;
+import io.streamnative.pulsar.handlers.mqtt.PacketIdGenerator;
 import io.streamnative.pulsar.handlers.mqtt.support.MessageBuilder;
 import java.io.IOException;
 import java.util.ArrayList;
@@ -67,7 +68,8 @@ public class PulsarMessageConverter {
         return MessageImpl.create(metadata, mqttMsg.payload().nioBuffer(), SCHEMA, topic.getName());
     }
 
-    public static List<MqttPublishMessage> toMqttMessages(String topicName, Entry entry, int messageId, MqttQoS qos) {
+    public static List<MqttPublishMessage> toMqttMessages(String topicName, Entry entry,
+                                                          PacketIdGenerator packetIdGenerator, MqttQoS qos) {
         ByteBuf metadataAndPayload = entry.getDataBuffer();
         MessageMetadata metadata = Commands.parseMessageMetadata(metadataAndPayload);
         if (metadata.hasNumMessagesInBatch()) {
@@ -81,7 +83,7 @@ public class PulsarMessageConverter {
                             single,
                             i, batchSize);
                     response.add(MessageBuilder.publish()
-                            .messageId(messageId)
+                            .messageId(packetIdGenerator.nextPacketId())
                             .payload(singleMessagePayload)
                             .topicName(topicName)
                             .qos(qos)
@@ -96,7 +98,7 @@ public class PulsarMessageConverter {
             }
         } else {
             return Collections.singletonList(MessageBuilder.publish()
-                    .messageId(messageId)
+                    .messageId(packetIdGenerator.nextPacketId())
                     .payload(metadataAndPayload)
                     .topicName(topicName)
                     .qos(qos)

--- a/tests/src/test/java/io/streamnative/pulsar/handlers/mqtt/mqtt5/hivemq/base/MQTT5BatchMessageTest.java
+++ b/tests/src/test/java/io/streamnative/pulsar/handlers/mqtt/mqtt5/hivemq/base/MQTT5BatchMessageTest.java
@@ -1,3 +1,16 @@
+/**
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 package io.streamnative.pulsar.handlers.mqtt.mqtt5.hivemq.base;
 
 import com.hivemq.client.mqtt.MqttGlobalPublishFilter;
@@ -5,12 +18,13 @@ import com.hivemq.client.mqtt.datatypes.MqttQos;
 import com.hivemq.client.mqtt.mqtt5.Mqtt5BlockingClient;
 import com.hivemq.client.mqtt.mqtt5.message.publish.Mqtt5Publish;
 import io.streamnative.pulsar.handlers.mqtt.base.MQTTTestBase;
-import org.apache.pulsar.client.api.Producer;
-import org.testng.Assert;
-import org.testng.annotations.Test;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.UUID;
+import org.apache.pulsar.client.api.Producer;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
 
 /**
  * The tests for MQTT protocol with batched messages.

--- a/tests/src/test/java/io/streamnative/pulsar/handlers/mqtt/mqtt5/hivemq/base/MQTT5BatchMessageTest.java
+++ b/tests/src/test/java/io/streamnative/pulsar/handlers/mqtt/mqtt5/hivemq/base/MQTT5BatchMessageTest.java
@@ -1,0 +1,54 @@
+package io.streamnative.pulsar.handlers.mqtt.mqtt5.hivemq.base;
+
+import com.hivemq.client.mqtt.MqttGlobalPublishFilter;
+import com.hivemq.client.mqtt.datatypes.MqttQos;
+import com.hivemq.client.mqtt.mqtt5.Mqtt5BlockingClient;
+import com.hivemq.client.mqtt.mqtt5.message.publish.Mqtt5Publish;
+import io.streamnative.pulsar.handlers.mqtt.base.MQTTTestBase;
+import org.apache.pulsar.client.api.Producer;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+/**
+ * The tests for MQTT protocol with batched messages.
+ * This feature is the not protocol version limit. So, it doesn't need to use an MQTT3 re-test.
+ */
+public class MQTT5BatchMessageTest extends MQTTTestBase {
+
+    @Test(timeOut = TIMEOUT)
+    public void testReceiveBatchMessageWithCorrectPacketId() throws Exception {
+        final String topic = "persistent://public/default/test-batch-message-1";
+        final Mqtt5BlockingClient client = MQTT5ClientUtils.createMqtt5Client(getMqttBrokerPortList().get(0));
+        client.connect();
+        client.subscribeWith()
+                .topicFilter(topic)
+                .qos(MqttQos.AT_LEAST_ONCE)
+                .send();
+        final Mqtt5BlockingClient.Mqtt5Publishes publishes = client.publishes(MqttGlobalPublishFilter.ALL, true);
+        Producer<byte[]> producer = pulsarClient.newProducer()
+                .enableBatching(true)
+                .batchingMaxMessages(5)
+                .topic(topic)
+                .create();
+        final List<String> payloads = new ArrayList<>();
+        for (int i = 0; i < 5; i++) {
+            final String payload = UUID.randomUUID().toString();
+            payloads.add(payload);
+            producer.sendAsync(payload.getBytes());
+        }
+        // The Hive MQ client doesn't have an API to get message packet id.
+        // However, when the Hive MQ client receives the duplicate packet id, it will throw an exception.
+        for (int i = 0; i < 5; i++) {
+            Mqtt5Publish message = publishes.receive();
+            message.acknowledge();
+            String payload = new String(message.getPayloadAsBytes());
+            Assert.assertTrue(payloads.contains(payload));
+        }
+        publishes.close();
+        client.disconnect();
+        producer.close();
+    }
+}

